src/services/job-processor/message-storage.ts

```diff
@@ -0,0 +1,110 @@
+import { ExitMessage, ExitMessageWithMetadata } from './service.js'
+
+export type ValidatorIndex = string
+export type TTL = number
+
+export class MessageStorage {
+  private MAX_TTL: TTL = 10
+
+  // Time to live counter for each message
+  private messageTTL: Map<ValidatorIndex, TTL> = new Map<ValidatorIndex, TTL>()
+
+  private messagesMetadatas: Map<ValidatorIndex, ExitMessageWithMetadata> =
+    new Map<ValidatorIndex, ExitMessageWithMetadata>()
+
+  private addOrUpdateMessage(
+    messageWithMetadata: ExitMessageWithMetadata
+  ): number {
+    const hasMsg = this.messagesMetadatas.has(
+      String(messageWithMetadata.data.message.validator_index)
+    )
+
+    this.messagesMetadatas.set(
+      String(messageWithMetadata.data.message.validator_index),
+      messageWithMetadata
+    )
+
+    this.setTTL(messageWithMetadata.data.message.validator_index, this.MAX_TTL)
+
+    if (!hasMsg) {
+      // msg was appended
+      return 1
+    }
+
+    // msg was updated (overwritten)
+    return 0
+  }
+
+  private setTTL(validatorIndex: ValidatorIndex, ttl: TTL) {
+    this.messageTTL.set(String(validatorIndex), ttl)
+  }
+
+  public updateMessages(messagesWithMetadata: ExitMessageWithMetadata[]): {
+    updated: number
+    added: number
+  } {
+    return messagesWithMetadata
+      .map((msg) => this.addOrUpdateMessage(msg))
+      .reduce(
+        (stats, x) => {
+          if (x === 1) {
+            stats.added++
+            return stats
+          }
+          stats.updated++
+          return stats
+        },
+        { updated: 0, added: 0 }
+      )
```

Comment:
It's quite hard to read and the value `1` is unclear. I would suggest to introduce enum for the result and counting the values with a simple filter.

Smth like this:
```ts
const processResults = messagesWithMetadata.map((msg) => this.addOrUpdateMessage(msg))
const updated = processResults.filter((result) => result === UPDATE_RESULT.updated).length
const added = processResults.filter((result) => result === UPDATE_RESULT.added).length

return { updated, added }
```